src/ts/components/sign-up/onboarding/index.tsx

```diff
@@ -0,0 +1,153 @@
+
+const RenderScreen = ({ ...rest }) => {
+  const dispatch = useDispatch();
+  const { screen, wallet, decodedAuthRequest, appManifest } = useSelector(
+    (state: IAppState) => ({
+      screen: selectCurrentScreen(state),
+      wallet: selectCurrentWallet(state),
+      decodedAuthRequest: selectDecodedAuthRequest(state),
+      appManifest: selectAppManifest(state),
+    })
+  );
+
+  // TODO
+  const doFinishSignIn = async () => {
+    if (!wallet || !appManifest || !decodedAuthRequest) {
+      console.log('Uh oh! Finished onboarding without auth info.');
+      return;
+    }
+    const gaiaUrl = 'https://hub.blockstack.org';
+    const authResponse = await wallet.identities[0].makeAuthResponse({
+      gaiaUrl,
+      appDomain: appManifest.start_url,
+      transitPublicKey: decodedAuthRequest.public_keys[0],
+    });
+    const redirect = `${decodedAuthRequest.redirect_uri}?authResponse=${authResponse}`;
+    window.open(redirect);
+    window.close();
+  };
+  const doFinishOnboarding = doFinishSignIn;
+
+  const [hasSaved, setHasSaved] = useState(false);
+  switch (screen) {
+    // intro / about
+    case Screen.INTRO:
+      return (
+        <Intro next={() => dispatch(doChangeScreen(Screen.CREATE))} {...rest} />
+      );
+
+    case Screen.HOW_IT_WORKS:
+      return (
+        <HowItWorks
+          back={() => dispatch(doChangeScreen(Screen.INTRO))}
+          {...rest}
+        />
+      );
+
+    // create
+    case Screen.CREATE:
+      return (
+        <Create
+          next={() => dispatch(doChangeScreen(Screen.SECRET_KEY))}
+          {...rest}
+        />
+      );
+
+    // Key screens
+    case Screen.SECRET_KEY:
+      return (
+        <SecretKey
+          next={() =>
+            dispatch(
+              doChangeScreen(hasSaved ? Screen.CONNECT_APP : Screen.SAVE_KEY)
+            )
+          }
+          {...rest}
+        />
+      );
+
+    case Screen.SAVE_KEY:
+      return (
+        <SaveKey
+          next={() => {
+            setHasSaved(true);
+            dispatch(doChangeScreen(Screen.CONNECT_APP));
+          }}
+          {...rest}
+        />
+      );
+
+    // Connect
+    case Screen.CONNECT_APP:
+      return (
+        <Connect
+          next={() => dispatch(doChangeScreen(Screen.CONNECTED))}
+          back={() => dispatch(doChangeScreen(Screen.SECRET_KEY))}
+          {...rest}
+        />
+      );
+
+    case Screen.CONNECTED:
+      return (
+        <Final
+          next={async () => {
+            await doFinishOnboarding();
+          }}
+          back={() => dispatch(doChangeScreen(Screen.SECRET_KEY))}
+          {...rest}
+        />
+      );
+
+    // Sign In
+
+    case Screen.SIGN_IN:
+      return (
+        <SignIn
+          next={async () => await doFinishSignIn()}
+          back={() => {
+            dispatch(doChangeScreen(Screen.SECRET_KEY));
+          }}
+          {...rest}
+        />
+      );
+
+    default:
+      return <Intro {...rest} />;
```

Comment:
Did we consider using a router for this? It would yield a better UX when in browser (navigation buttons would work as well), and afaik, react-router works in an extension as well.

Reply:
It's probably a good thing to consider earlier than later. Adding a router to onboarding does add complexity (i.e. what if the user ends up on this middle step's route, but with no auth information?), but can add convenience to the user and the developer. We also are already building out multiple different HTML pages to handle the different pages an extension has (popup, browser action, home page). The way extensions work, we might need to keep separate files, but we can definitely do hash-based routing.